python/ray/util/concurrent/futures/ray_executor.py

```diff
@@ -0,0 +1,658 @@
+        None, RayExecutor will default to shutting down the Ray instance only
+        if it was initialised during instantiation of the current RayExecutor,
+        otherwise it will not be shutdown.
+    initializer : Callable[..., Any] | None
+        A function that will be called remotely in the actor context before the
+        submitted task (for compatibility with
+        concurrent.futures.ThreadPoolExecutor).
+    initargs : tuple[Any, ...] | None
+        Arguments for initializer (for compatibility with
+        concurrent.futures.ThreadPoolExecutor).
+    max_tasks_per_child : int | None
+        The maximum number of tasks to be performed by an actor before it is
+        gracefully killed and replaced (for compatibility with
+        concurrent.futures.ProcessPoolExecutor).
+    mp_context : Any | None
+        This is only included for compatibility with
+        concurrent.futures.ProcessPoolExecutor but is unused.
+    actor_pool_type: str | None
+        The type of actor pool to use: either 'random' (which will assign tasks
+        to randomly selected actors), or 'roundrobin' (which will simply assign
+        tasks to actors sequentially.
+
+    All additional keyword arguments are passed to ray.init()
+    (see https://docs.ray.io/en/latest/ray-core/package-ref.html#ray-init).
+
+    For example, this will connect to a cluster at the specified address:
+    .. testcode::
+
+        RayExecutor(address='192.168.0.123:25001')
+
+    Note: excluding an address will initialise a local Ray cluster.
+    """
+
+    #: Aggregated Futures from initiated tasks.
+    futures: list[Future[Any]]
+    #: A container of Actor objects over which the tasks will be distributed.
+    actor_pool: _AbstractActorPool
+
+    def __init__(
+        self,
+        max_workers: Optional[int] = None,
+        shutdown_ray: Optional[bool] = None,
+        initializer: Optional[Callable[..., Any]] = None,
+        initargs: tuple[Any, ...] = (),
+        max_tasks_per_child: Optional[int] = None,
+        mp_context: Optional[Any] = None,
+        actor_pool_type: str = "random",
+        **kwargs: Any,
+    ):
+
+        """RayExecutor handles existing Ray instances and shutting
+        down as follows:
+
+            1. If an existing cluster is discovered in the current scope,
+            RayExecutor will simply connect to this cluster. By default, this
+            cluster will not be destroyed when RayExecutor.shutdown() is
+            called.
+
+            2. If no existing cluster is discovered in the current scope,
+            RayExecutor will instantiate a new Ray cluster instance. By
+            default, this cluster will be destroyed when RayExecutor.shutdown()
+            is called.
+
+            3. If RayExecutor.shutdown_ray is set to True or False, this will
+            override the behaviour above.
+        """
+        self._shutdown_lock: bool = False
```

Comment:
This is not a lock. Did you mean to create an `RLock` here? If not, then rename it appropriately.